docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+  - `send loop <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L381-L402>`_
+- Related RFCs 
+  - `RFC 002: IPC Ecosystem <./rfc-002-ipc-ecosystem.md>`_ 
+  - `RFC 003: Performance Questions <./rfc-003-performance-questions.md>`_ 
+
+Discussion
+----------
+
+Changes to Published Events
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of this process, the Tendermint team should do a study of the existing
+event types and ensure that there are viable production use cases for
+subscriptions to all event types. Instinctively it seems plausible that some
+of the events may not be useable outside of tendermint, (e.g. ``TimeoutWait``
+or ``NewRoundStep``) and it might make sense to remove them. Certainly, it
+would be good to make sure that we don't maintain infrastructure for unused or
+un-useful message indefinitely.
+
+Blocking Subscription
+~~~~~~~~~~~~~~~~~~~~~
+
+The blocking subscription mechanism makes it possible to have *send*
+operations into the subscription channel be un-buffered (the event processing
+channel is still buffered.) In the blocking case, events from one subscription
+can block processing that event for other non-blocking subscriptions. The main
+case, it seems for blocking subscriptions is ensuring that a transaction has
+been committed to a block for ``BroadcastTxCommit``. Removing blocking
+subscriptions entirely, and potentially finding another way to implement
+``BroadcastTxCommit``, could lead to important simplifications and
+improvements to throughput without requiring large changes.
+
+Subscription Identification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before `#6386 <https://github.com/tendermint/tendermint/pull/6386>`_, all
+subscriptions were identified by the combination of a client ID and a query,
+and with that change, all it became possible to identify subscription's based
+on an ID but compatibility with the legacy identification means that there's a
+good deal of legacy code as well as client side efficiency that could be
+improved. 
+
+Pubsub Changes
+~~~~~~~~~~~~~~
+
+The pubsub core should be implemented in a way that removes the possibility of
+backpressure from the event system to impact the core system *or* for one
+subscription to impact the behavior of another area of the
+system. Additionally, because the current system is implemented entirely in
+terms of a collection of buffered channels, the event system (and large
+numbers of subscriptions) can be a source of memory pressure. 
+
+These changes could include: 
+
+- explicit cancellation and timeouts promulgated from callers (e.g. RPC end
+  points, etc,) this should be done using contexts.
+
+- subscription system should be able to spill to disk to avoid putting memory
+  pressure on the core behavior of the node (consensus, gossip).
+  
+- subscriptions implemented as cursors rather than channels, with either
+  condition variables to simulate the existing "push" API or a client side
```

Comment:
I'd be a little hesitant to perform to much of a large scale rebuild of the internal system until we have a better understanding of the throughput. At some level, we have a system that behaves like a big queue with multiple readers. I definitely think we should avoid letting readers starve each other out in the way we have it now, but I also think that if you have a big queue and your throughput isn't good enough, eventually you're going to back up. I similarly worry about this with the spilling to disk case. If we have to start spilling to disk because a reader is too slow, maybe we don't ever stop spilling to disk?

Reply:
I agree—this is part of what I was getting at above, about defining what capabilities we want the system to provide. I think a lot of these things could be improved without rewriting from scratch—the API isn't that complex, and there's a lot of low-hanging free improvements that wouldn't necessarily break other components. 

I _do_ think we should define the constraints, though, and use that to do larger changes. Event subscription seems to be important enough for a lot of users that IMO it's worth some investment.

Reply:
> Event subscription seems to be important enough for a lot of users that IMO it's worth some investment.

The importance of event subscription has certainly evolved over the last year. I think prior to the introduction of relayers, the events was just a nice feature that some clients might have used via a websocket. It certainly wasn't designed with strong reliability/robustness guarantees in mind.

I tend to agree that the more immediate challenge is distilling what items can be resolved now (and in a non-breaking way) and what items need to be pushed back for a more large scale redesign